File: C++/Sort_Bubble.cpp
/*
 * C++ PROGRAMMING EXAMPLES EXAMPLES
 * 
 * This file demonstrates C++ programming concepts and techniques:
 * - Core concepts and syntax
 * - Practical examples and usage patterns
 * - Best practices and optimization techniques
 * - Real-world applications and scenarios
 * 
 * Learning Objectives:
 * - Understand fundamental C++ PROGRAMMING concepts
 * - Learn proper syntax and implementation
 * - Master different techniques and approaches
 * - Apply knowledge in practical scenarios
 * - Develop problem-solving skills
 * 
 * Prerequisites:
 * - Basic C++ knowledge
 * - Understanding of data types and variables
 * - Familiarity with control structures
 * - Basic understanding of functions and classes
 * 
 * Key Topics Covered:
 * - Syntax and language features
 * - Implementation techniques
 * - Performance considerations
 * - Error handling and best practices
 * - Real-world applications
 */

/* Bubble Sort 
Time Complexity of Bubble Sort
Case	Comparisons/Swaps	Time Complexity
Best Case	No swaps needed (already sorted)	O(n)
Average Case	Regular comparisons/swaps	O(n²)
Worst Case	Reverse sorted array	O(n²)

Space Complexity
O(1) – Bubble sort is an in-place sorting algorithm (uses no extra space except a few variables).

https://visualgo.net/en/sorting
*/

#include <iostream>
using namespace std;

void bubbleSort(int arr[], int n) {
    bool swapped;

    for (int i = 0; i < n - 1; i++) {
        swapped = false;

        // Last i elements are already sorted
        for (int j = 0; j < n  - i - 1; j++) {
            // Swap if current element is greater than the next
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                swapped = true;
            }
        }

        // If no elements were swapped, the array is sorted
        if (!swapped)
            break;
    }
}

void printArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
    cout << endl;
}

int main() {
    int arr[] = {64, 34, 25, 12, 22, 11, 90};
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Original array: ";
    printArray(arr, n);

    bubbleSort(arr, n);

    cout << "Sorted array: ";
    printArray(arr, n);

    return 0;
}

File: python_old/sort_bubble.py
'''
Bubble Sort 
Time Complexity of Bubble Sort
Case    Comparisons/Swaps   Time Complexity
Best Case   No swaps needed (already sorted)    O(n)
Average Case    Regular comparisons/swaps   O(n²)
Worst Case  Reverse sorted array    O(n²)

Space Complexity
O(1) Bubble sort is an in-place sorting algorithm (uses no extra space except a few variables).

https://visualgo.net/en/sorting
'''

def bubble_sort(arr):
    n = len(arr)
    for i in range(n):
        # Track whether any swap was made in this pass
        swapped = False
        for j in range(0, n - i - 1):  # Last i elements are already sorted
            if arr[j] > arr[j + 1]:
                # Swap if the current item is greater than the next
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        # If no swaps were made, the list is already sorted
        if not swapped:
            break

# Example usage:
arr = [64, 34, 25, 12, 22, 11, 1, 23, 7, 5, 9]
print("Original array:", arr)
bubble_sort(arr)
print("Sorted array:  ", arr)
